handlers/proxy_handler.cc

```diff
@@ -0,0 +1,240 @@
+	}
+}
+
+void ProxyHandler::ConnectSocketToEndpoint(tcp::socket* socket, std::string host, std::string port) {
+	tcp::resolver resolver(io_service_);
+	tcp::resolver::query query(host, port);
+	tcp::resolver::iterator endpoint_iterator = resolver.resolve(query);
+	boost::asio::connect(*socket, endpoint_iterator);
+}
+
+void ProxyHandler::WriteToSocket(tcp::socket* socket, const std::string& requestStr) {
+	boost::asio::streambuf reqBuf;
+	std::ostream request_stream(&reqBuf);
+	request_stream << requestStr;
+	boost::asio::write(*socket, reqBuf);
+}
+
+boost::system::error_code ProxyHandler::SocketReadToEOF(tcp::socket* socket, boost::asio::streambuf* buf, std::string* data) {
+	boost::system::error_code error;
+	while (true) {
+		boost::asio::read(*socket, *buf, boost::asio::transfer_at_least(1), error);
+		if (buf->size() == 0) {
+			break;
+		}
+		auto bufs = buf->data();
+		std::string bodyChunk(boost::asio::buffers_begin(bufs), boost::asio::buffers_begin(bufs) + buf->size());
+		*data += bodyChunk;
+		buf->consume(buf->size());
+	}
+	return error;
+}
+
+boost::system::error_code ProxyHandler::SocketReadUntil(
+		tcp::socket* socket, boost::asio::streambuf* buf, const std::string& sep) {
+	boost::system::error_code error;
+	boost::asio::read_until(*socket, *buf, sep, error);
+	if (error && error != boost::asio::error::eof) {
+		printf("ProxyHandler: error! %s\n", error.message().c_str());
+	}
+	return error;
+}
+
+bool ProxyHandler::ReadNextHeader(std::istream* response_stream, std::string* header_key, std::string* header_value) {
+	std::string header;
+	if ( ! std::getline(*response_stream, header)) return false;
+	if (header == "\r") {
+		return false; // blank line = end of headers
+	}
+	auto colonPos = header.find(':');
+	if (colonPos == std::string::npos) {
+		printf("ProxyHandler: warning: malformed header '%s'\n", header.c_str());
+		*header_key = "";
+		*header_value = "";
+		return true; // possibly a transient error, try to get further headers.
+	}
+	*header_key = header.substr(0, colonPos);
+	*header_value = boost::algorithm::trim_copy(header.substr(colonPos+1));
+	return true;
+}
+
+RequestHandler::Status ProxyHandler::SendRequestToServer(
+		const std::string& host, const std::string& port,
+		const Request& request, Response* response, int depth) {
+
+	if (depth <= 0) {
+		printf("ProxyHandler: Request redirect depth limit reached. Aborting.\n");
+		response->SetStatus(Response::NOT_FOUND);
+		return OK;
+	}
+
+	printf("ProxyHandler: connecting to %s (port %s)\n", host.c_str(), port.c_str());
+	try {
+		tcp::socket socket(io_service_);
+
+		ConnectSocketToEndpoint(&socket, host, port);
+
+		boost::asio::streambuf reqBuf;
+		std::ostream request_stream(&reqBuf);
+
+		MutableRequest modified_request(request);
+		ModifyRequestForProxy(request, &modified_request);
+
+		WriteToSocket(&socket, modified_request.ToString());
+
+		const size_t bufSize = 8192; // guaranteed large enough for HTTP status line
```

Comment:
luckily boost will not overwrite beyond the end of the buffer, it will just stop reading input. So while it's not exactly well defined, I don't think it's a stack-smash vulnerability per-se.